File: ros2_ws/src/py_tf2/py_tf2/tf2_listener.py
from typing import Optional

import math
import rclpy
from rclpy.node import Node
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener
from tf2_ros import LookupException, ConnectivityException, ExtrapolationException
from rclpy.time import Time


class FrameListener(Node):
    def __init__(self) -> None:
        super().__init__("frame_listener")

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        self.timer = self.create_timer(1.0, self.timer_callback)

    def timer_callback(self) -> None:
        try:
            t = self.tf_buffer.lookup_transform("base_link", "moving_frame", Time())

            translation = t.transform.translation
            rotation = t.transform.rotation

            distance = math.sqrt(translation.x**2 + translation.y**2 + translation.z**2)

            self.get_logger().info(
                f"moving_frame at ({translation.x:.3f}, {translation.y:.3f}, {translation.z:.3f}), "
                f"distance from base_link = {distance:.3f} m, "
                f"rotation = ({rotation.x:.3f}, {rotation.y:.3f}, {rotation.z:.3f}, {rotation.w:.3f})"
            )

        except (LookupException, ConnectivityException, ExtrapolationException) as e:
            self.get_logger().warn(f"Could not transform: {e}")


def main(args: Optional[list[str]] = None) -> None:
    rclpy.init(args=args)
    node = FrameListener()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2_ws/src/cpp_tf2/src/tf2_listener.cpp
#include "cpp_tf2/tf2_listener.hpp"

#include <chrono>

namespace learning_ros2::cpp_tf2 {

FrameListener::FrameListener(const rclcpp::NodeOptions& options)
    : rclcpp::Node("frame_listener", options) {
  tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

  timer_ = this->create_wall_timer(
      std::chrono::seconds(1), std::bind(&FrameListener::timer_callback, this));
}

void FrameListener::timer_callback() {
  geometry_msgs::msg::TransformStamped t;

  try {
    t = tf_buffer_->lookupTransform("base_link", "moving_frame",
                                    tf2::TimePointZero);
  } catch (const tf2::TransformException& ex) {
    RCLCPP_WARN(this->get_logger(), "Could not transform: %s", ex.what());
    return;
  }

  const auto& translation = t.transform.translation;
  const auto& rotation = t.transform.rotation;

  double distance =
      std::sqrt(translation.x * translation.x + translation.y * translation.y +
                translation.z * translation.z);

  RCLCPP_INFO(this->get_logger(),
              "moving_frame at (%.3f, %.3f, %.3f), "
              "distance from base_link = %.3f m, "
              "rotation = (%.3f, %.3f, %.3f, %.3f)",
              translation.x, translation.y, translation.z, distance, rotation.x,
              rotation.y, rotation.z, rotation.w);
}

}  // namespace learning_ros2::cpp_tf2
